range_queries/heavy_light_decomposition.cpp

```diff
@@ -0,0 +1,578 @@
+/**
+ * @file
+ * @brief [Heavy Light Decomposition](https://en.wikipedia.org/wiki/Heavy_path_decomposition) implementation
+ * @author: [Aniruthan R](https://github.com/aneee004)
+ *
+ * @details
+ * Function Documentation: 
+ * Heavy-Light Decomposition is a technique on trees, that supports the following:
+ * 1: Update node s, with a value v
+ * 2: Return the (sum) of all node values on the simple path from a to b
+ * (sum) can also be replced with XOR, OR, AND, min, or max
+ *
+ * The update is done in O(log n) time, and
+ * the query is done in O(log^2 n) time with HLD
+ * where, n is the number of nodes
+ *
+ * The template type is the data type of the value stored in the nodes.
+ * If a non-primitive data-type is used as a template, 
+ * the coressponding operators must be overloaded.
+ *
+ * An HLD object can only be created with a constant number of nodes, and
+ * it cannot be changed later. Creaty an empty instance is not supported.
+ *
+ * To start answering updates and queries,
+ * 1. Create an instance of HLD<X> object (obj), with the required data type.
+ * 2. Read in the edge/parent information and update it with obj.add_edge().
+ * Note: The edges addes must be 0 indexed.
+ * 3. Create a vector with initial node values, and call set_node_val() with it.
+ * 4. Call obj.init() to populate the required information for supporting operations.
+ * 5. Call obj.update(node, new_val), to update the value at index 'node' to the new value.
+ * Note: node must be 0 indexed
+ * 6. Call obj.query(a, b) to get the (sum) of node values in the simple path from a to b.
+ * Note: a and b, must be 0 indexed.
+ *
+ * Sample I/O at the bottom.
+ * TODO: Support edge weight queries, by storing the edge weight value in it's child
+ * algorithm verified by testing in CSES path queries: https://cses.fi/problemset/task/1138
+ */
+
+#include <iostream>
+#include <vector>
+#include <cstring>
+#include <string>
+#include <list>
+#include <cmath>
+#include <cassert>
+#include <algorithm>
+#include <numeric>
+
+/**
+ * @namespace range_queries
+ * @brief Algorithms and Data Structures that support range queries and updates.
+*/
+namespace range_queries {
+namespace heavy_light_decomposition {
+
```

Comment:
Please remove this line.

Reply:
The namespace `heavy_light_decomposition` ?

Reply:
No, the line 56. 🙂

Reply:
Oh the blank line xD. Alright.